contrib/mod_tls.c

```diff
@@ -13367,32 +13392,64 @@ MODRET set_tlscertchain(cmd_rec *cmd) {
   return PR_HANDLED(cmd);
 }
 
-/* usage: TLSCipherSuite string */
+/* usage: TLSCipherSuite [protocol] string */
 MODRET set_tlsciphersuite(cmd_rec *cmd) {
   config_rec *c = NULL;
   char *ciphersuite = NULL;
+  int protocol = 0;
   SSL_CTX *ctx;
 
   CHECK_ARGS(cmd, 1);
   CHECK_CONF(cmd, CONF_ROOT|CONF_VIRTUAL|CONF_GLOBAL);
 
-  ciphersuite = cmd->argv[1];
-  c = add_config_param(cmd->argv[0], 1, NULL);
+  if (cmd->argc-1 == 1) {
+    ciphersuite = cmd->argv[1];
 
-  /* Make sure that EXPORT ciphers cannot be used, per Bug#4163. Note that
-   * this breaks system profiles, so handle them specially.
-   */
-  if (strncmp(ciphersuite, "PROFILE=", 8) == 0) {
-    ciphersuite = pstrdup(c->pool, ciphersuite);
+  } else if (cmd->argc-1 == 2) {
+    char *protocol_text;
 
-  } else {
-    ciphersuite = pstrcat(c->pool, "!EXPORT:", ciphersuite, NULL);
+    protocol_text = cmd->argv[1];
+    if (strcasecmp(protocol_text, "TLSv1.3") == 0) {
+      protocol = TLS_PROTO_TLS_V1_3;
+
+    } else {
+      CONF_ERROR(cmd, pstrcat(cmd->tmp_pool,
+        "unknown/unsupported protocol specifier: ", protocol_text, NULL));
+    }
+
+    ciphersuite = cmd->argv[2];
+  }
+
+  c = add_config_param(cmd->argv[0], 2, NULL, NULL);
+
+  if (protocol != TLS_PROTO_TLS_V1_3) {
```

Comment:
This block pstrdup()s ciphersuite, but only in the !TLSv1.3 case. It needs to pstrdup() in both cases, like this:
```
  if (protocol == TLS_PROTO_TLS_V1_3) {
    ciphersuite = pstrdup(c->pool, ciphersuite);

  } else {
    if (strncmp(ciphersuite, "PROFILE=", 8) == 0) {
      ciphersuite = pstrdup(c->pool, ciphersuite);

    } else {
      ciphersuite = pstrcat(c->pool, "!EXPORT:", ciphersuite, NULL);
    }
  }
```

Reply:
Ah, yes.  Good catch!  Now fixed.